### ISSUE TITLE
test_batch_norm_grad_op: invalid negative variance input masked by hardware FP behavior, now failing after tt-metal uplift

### ISSUE THREAD

**Issue by hkwonTT** [tenstorrent/tt-mlir#8844]
## Summary

`test_batch_norm_grad_op` was generating **negative variance** values via `torch.randn`, which is mathematically invalid for batch normalization. This bug was silently hidden by a specific (and incorrect per IEEE 754) hardware floating-point behavior on Wormhole. A tt-metal accuracy improvement changed that hardware behavior, unmasking the test defect and causing the test to fail.

## Failing Test

```
FAILED test/python/golden/test_stablehlo_ops.py::test_batch_norm_grad_op[module_batch_norm_grad-ttnn]
TTBuilderGoldenException: Failed: program-level output golden comparison failed,
actual_pcc=0.023481420220723564 < expected_pcc=0.99
```

Reported in CI run: https://github.com/tenstorrent/tt-mlir/actions/runs/27785314719/job/82224348493?pr=8840

### 1. Invalid test input: negative variance

The `module_batch_norm_grad` function used `@builder.func` with `torch.randn` to auto-generate all inputs, including `variance` (shape `[64]`, dtype `float32`). Since `torch.randn` samples from a standard normal distribution, roughly half the generated variance values are negative.

**Variance in batch normalization represents `E[(x - mean)^2]` -- it is always non-negative by definition.** Passing negative variance to `batch_norm_grad` is mathematically invalid.

### 2. Why this passed before

The `stablehlo.batch_norm_grad` lowering computes (simplified):

```
stddev     = sqrt(variance + epsilon)            # sqrt(negative) = NaN  [on CPU and device]
normalized = centered_operand / stddev           # x / NaN              [behavior differs]
grad_scale = sum(grad_output * normalized)       # sum(...)
```

On **Wormhole hardware (before tt-metal commit `2d254d5b38e`)**, the SFPU division computed:

```
recip(NaN) ~= 0.0   =>   x * 0.0 = 0.0
```

So `x / NaN` produced `0.0` on the device (incorrect per IEEE 754, but effectively suppressed NaN propagation). Then `sum(0)` produced `0.0`. Meanwhile, the CPU golden's `get_pcc()` masks NaN/Inf to `0` before computing PCC. Both sides landed on `~0`, PCC was high, and the test passed -- despite invalid inputs.

### 3. What changed: tt-metal commit `2d254d5b38e`

[`"Improve ttnn.logit Accuracy (#46471)"`](https://github.com/tenstorrent/tt-metal/commit/2d254d5b38e42963b73be4e2e47551676a03f246) added a Markstein residual correction to `calculate_sfpu_binary_div` in `ckernel_sfpu_binary.h` for FP32 accuracy:

```cpp
// Before
sfpi::vFloat result = in0 * sfpu_reciprocal_iter<2>(in1);

// After: Markstein refinement
sfpi::vFloat r = sfpu_reciprocal_iter<2>(in1);
sfpi::vFloat result = in0 * r;
// residual correction: e = in0 - result * in1
// result = result + e * r
```

This is **correct per IEEE 754**. As a result, `x / NaN` now propagates `NaN` instead of silently returning `0.0`.

With NaN now propagating through division, `ttnn.sum` receives NaN-containing inputs and returns `inf` (or large finite values) -- creating a large mismatch against the CPU golden (which masks NaN to `0`).

### 4. Causal chain (before vs. after)

| Step | Before uplift | After uplift |
|---|---|---|
| `variance` | negative (from `randn`) | negative (same) |
| `sqrt(variance + eps)` | `NaN` | `NaN` |
| `x / NaN` on device | `0.0` (hidden NaN) | `NaN` (correct) |
| `sum(...)` on device | `0.0` | `inf` |
| CPU golden PCC | NaN masked to 0, PCC high | `0` vs `inf`, PCC ~0.02 |
| Test result | PASS (incorrectly) | FAIL |

## Bisection Evidence

Confirmed by checking out the tt-metal submodule kernel files at:
- `349db7f58aa` (just before the logit commit) + run test --> **PASSED**
- `2d254d5b38e` (logit commit applied) + run test --> **FAILED** with `pcc=0.023`

## Fix

Override the variance golden in `module_batch_norm_grad` to use `torch.rand` (values in `[0, 1)`) instead of the auto-generated `torch.randn`:

```python
# variance must be non-negative (it represents E[(x-mean)^2]).
# torch.randn can produce negative values causing sqrt(negative) = NaN.
builder.set_goldens(
    {variance: torch.rand(builder.get_shape(variance), dtype=torch.float32)}
)
```

This fix is included in PR #8840.

## Notes

- The tt-metal logit accuracy fix (`#46471`) is **correct** and not the problem.
- The test was always wrong; the hardware behavior change simply exposed it.
- A related latent issue in `ttnn.sum` (returns `inf` instead of `NaN` for NaN inputs, violating IEEE 754) was observed but is a separate tt-metal concern.

**Comment by hkwonTT** [comment]
@jgrimTT FYI

**Comment by hkwonTT** [comment]
Fixed with #8840
